src/Microsoft.Azure.CosmosRepository/Repositories/DefaultRepository.Paging.cs

```diff
@@ -102,5 +103,44 @@ internal sealed partial class DefaultRepository<TItem>
                 charge + countResponse?.RequestCharge ?? 0,
                 resultingContinationToken /* This was missing, is this correct? */);
         }
+
+        /// <inheritdoc/>
+        public async IAsyncEnumerable<(TItem Item, int? Total)> PageAsync(
+            Expression<Func<TItem, bool>>? predicate = null,
+            int? maxNumber = null,
+            bool returnTotal = false,
+            [EnumeratorCancellation]
+            CancellationToken cancellationToken = default)
+        {
+            Container container = await _containerProvider.GetContainerAsync()
+                .ConfigureAwait(false);
+
+            IQueryable<TItem> query = container
+                .GetItemLinqQueryable<TItem>()
+                .Where(_repositoryExpressionProvider
+                    .Build(predicate ?? _repositoryExpressionProvider.Default<TItem>()));
+
+            int? total = null;
+            if (returnTotal)
+            {
+                total = (await query.CountAsync(cancellationToken))?.Resource;
+            }
+
+            query = maxNumber is int max and > 0 ? query.Take(max) : query;
+
+            _logger.LogQueryConstructed(query);
+
+            IEnumerable<TItem>? items = await GetByQueryAsync(

```

Comment:
Would this implementation, using a non-async enumerable approach to load all the records let's say 10 pages into memory, before then starting to stream them back to the client? Would we want to stream the first page back to the caller as soon as it's available to us?

Reply:
Yeah, that is a really good callout. I was trying to think of a more appropriate approach. Perhaps, I should refactor this to:

- Get the count first.
- Using the count determine a reasonable page size.
- Page over the data, then for each page yield it back.

That's probably better, right? Or do you have another idea?

Reply:
Thoughts on this:
https://github.com/IEvangelist/azure-cosmos-dotnet-repository/pull/294/files#diff-7e1d408ab862dfcc50eebd4168a676306ab6f4c74ef877d6527375f6b7c688dfR129-R147

Reply:
Yeah that could work, but I am pretty sure running a count query across a large data set can rack up a hefty RU bill, @robertbennett1998 did you have the docs or article we read around this?

I am wondering whether we allow the caller to specify there chunk size, we can assume a reasonable default maybe?

Reply:
I'm asking one of the PMs on the team, I'm not sure if there is any way to avoid this. Perhaps we just omit the count?

Reply:
Could the continuation token being null let you know there are no more pages?

Reply:
Any news on this?

Reply:
I was trying to ask this question to @jcocchi. Let's ask her here, and see if this additional context helps to make the question more clear.